### DISCUSSION TITLE
Why i'm not seeing debug log?

### DISCUSSION THREAD

**Discussion by Andrew-Garage** [eModbus/eModbus#264] [Q&A] [answered]
Hi, everyone!
I'm trying to figure out why im not seeing debug messages in "verbose mode". i suppose that i should see everyone of them. Here is my code: 
[[](url)](https://github.com/Andrew-Garage/Laboratory/blob/821a26fd2e614ac69c3f5834658ac8a3a8fc0760/TEST_%20Modbus&7SegDisp.ino)
The code is raw, but works well, I debugged it without logs, using "serial print". But to continue, I think I need to figure out what I was doing wrong.

This is a log from serial monitor after reset. as you can see, some logs are all present.

```
11:36:53.043 -> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
11:36:53.043 -> configsip: 0, SPIWP:0xee
11:36:53.043 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
11:36:53.043 -> mode:DIO, clock div:1
11:36:53.043 -> load:0x3fff0030,len:1344
11:36:53.043 -> load:0x40078000,len:13836
11:36:53.043 -> load:0x40080400,len:3608
11:36:53.089 -> entry 0x400805f0
11:36:53.197 -> [    22][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
11:36:53.244 -> __ Serial OK __
11:36:53.552 -> [   351][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=21 scl=22 freq=100000
11:37:11.031 -> [1;33m[E] 17830| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E5 - Packet length error
11:37:11.031 -> [0m[1;33m[E] 17836| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E5 - Packet length error
11:37:11.031 -> [0m[1;33m[E] 20350| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E2 - CRC check error
11:37:13.523 -> [0m[1;33m[E] 20352| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E5 - Packet length error
11:37:13.569 -> [0m[1;33m[E] 48992| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E2 - CRC check error
11:37:42.164 -> [0m[1;33m[E] 48996| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E5 - Packet length error
11:37:42.210 -> [0m[1;33m[E] 51717| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E2 - CRC check error
11:37:44.921 -> [0m[1;33m[E] 51720| ModbusServerRTU.cpp  [ 254] serve: RTU receive: E5 - Packet length error
11:37:44.921 -> [0m
```

E2 and E5 is known problem, I'll come back to this a bit later. As i understand, there is at least one message from ModbusServerRTU.h missed:
`LOG_D("Server task %d started. Interval=%d\n", (uint32_t)serverTask, MSRinterval);LOG_D("Server task %d started. Interval=%d\n", (uint32_t)serverTask, MSRinterval);`
Am i right?

**Comment by Miq1** [comment, marked as answer]
You will need to compile all with the LOG_LEVEL_VERBOSE setting, including the eModbus library. Another thing you could try is to set the actual log level at runtime by ``MBUlogLvl = LOG_LEVEL_VERBOSE;`` in your setup, in case it just is set to ERROR still.

**Reply by Andrew-Garage** [reply]
I've added this string `Serial.println(MBUlogLvl);` in one of my tasks and this string before 
  ```
  MBUlogLvl = LOG_LEVEL_VERBOSE;
  MBserver.start(); 
```
as a result, you can see 2(ERROR) switching to 6(VERBOSE):
```
13:56:13.092 -> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
13:56:13.138 -> configsip: 0, SPIWP:0xee
13:56:13.138 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
13:56:13.138 -> mode:DIO, clock div:1
13:56:13.138 -> load:0x3fff0030,len:1344
13:56:13.138 -> load:0x40078000,len:13836
13:56:13.138 -> load:0x40080400,len:3608
13:56:13.138 -> entry 0x400805f0
13:56:13.294 -> [    22][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
13:56:13.294 -> __ Serial OK __
13:56:13.606 -> [   351][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=21 scl=22 freq=100000
13:56:13.730 -> 2
13:56:14.727 -> 6
13:56:15.722 -> 6
13:56:16.726 -> 6
13:56:17.707 -> 6
```
But no logging is happens) After that i removed MBUlogLvl = LOG_LEVEL_VERBOSE; from setup, to try your first offer.

You say "to to compile all the eModbus library". What did you mean? I'm beginner in this. I've added all header-files from C:\Users\PC\Documents\Arduino\libraries\eModbus\src , and got no logging (it's stay 2-ERROR still)

```
#include <Arduino.h>
#include <HardwareSerial.h>

#include <CoilData.h>
#include <ModbusBridgeEthernet.h>
#include <ModbusBridgeRTU.h>
#include <ModbusBridgeTemp.h>
#include <ModbusBridgeWiFi.h>
#include <ModbusClient.h>
#include <ModbusClientRTU.h>
#include <ModbusClientTCP.h>
#include <ModbusClientTCPasync.h>
#include <ModbusError.h>
#include <ModbusMessage.h>
#include <ModbusTypeDefs.h>
#include <ModbusServer.h>
#include <ModbusServerRTU.h>
#include <ModbusServerEthernet.h>
#include <ModbusServerTCPasync.h>
#include <ModbusServerTCPtemp.h>
#include <ModbusServerWiFi.h>
#include <options.h>
#include <RTUutils.h>

```
After that i tried to change log-mode directly in Logging.h:
```
#ifndef LOG_LEVEL
#define LOG_LEVEL LOG_LEVEL_ERROR  //  LOG_LEVEL_VERBOSE (I used this)
#endif

#ifndef LOCAL_LOG_LEVEL
#define LOCAL_LOG_LEVEL LOG_LEVEL
#endif
```
And it finally starts to log as it should)) That comrad #189 use almost the same code as i, but he got logging. What's the diffirence? I'm missing something

**Reply by Miq1** [reply]
Are you using the Arduino IDE? Then it may be easier for you to modify the default log level in the ``Logging.h`` file. Normally this would be done by a compiler flag, but that is slightly difficult in Arduino IDE as compared to PlatformIO, where you can add it in ``platformio.ini``.

**Reply by Andrew-Garage** [reply]
Yes, Arduino IDE 2.03. This is sad, because this library is positioned as "Key features: for use in the Arduino framework", and I have spent a week studying it, and I plan to add a can bus to the project, and this is not a homeproject, but a project for the company :)) 
It would be great to change the log level quickly and in those parts of the code, where it is really necessary, at least during debugging or when the device is under repair. 
What should I do/read to fix logging? If I understood correctly, Do I need to dig into the ESP32 core for arduino? Maybe there is an easier way?

**Reply by Miq1** [reply]
As I wrote: locate the ``Logging.h`` file in the library path of the Arduino IDE, open it and change line 7 from
```
#define LOG_LEVEL LOG_LEVEL_ERROR
```
into
```
#define LOG_LEVEL LOG_LEVEL_VERBOSE
```
Then recompile your project. The log level now should be verbose as a default. You may have to change it back to error for a product to be deployed to reduce the output to the necessary.

If you are doing professional development I only can advise to use PlatformIO instead of the Arduino IDE. It is much more flexible and consistent IMHO. There you would have to add a ``build_flags = -DLOG_LEVEL=6`` line to ``platformio.ini`` and you are done.

**Reply by Andrew-Garage** [reply]
Thanks for reply and help! I agree with you about IDE, this is the more correct way. But, is it possible to expect that this feature will appear in the next versions?

**Reply by Miq1** [reply]
Erm... What feature are you referring to? The verbose logging? That completely does depend on your compilation, there is nothing to be modified in the library. If we left it on verbose by default, everyone would have tons of output on the serial that is necessary only in exceptional cases. Additionally, the lib will stay smaller if less debug output is compiled in.

**Reply by Andrew-Garage** [reply]
I'm reffering on this part of documentation:
[](https://emodbus.github.io/logging)
```
Put #include "Logging.h" in your source file.
#define LOG_LEVEL LOG_LEVEL_... before including Logging.h will set the global log level for the compiler. 
```
You said that it doesn't work for me, because i'm using arduino ide. so i'm asking, will it work on Arduino IDE ever?)) . Because that API much more convienent than a constant need "to upgrade" Logging.h )

**Reply by Miq1** [reply]
Ah, we talked of different things a bit, so it seems.

The documentation is referring to changing the log level of any single source file differently. If you put the lines as shown in one of your sources, log calls **within that file** will follow the rule. This is done on purpose to allow to tailor the verbosity for the spot you are currently interested in and not have all cluttered by the log output of less interesting parts of the code.

The global log level can only be adjusted by
- setting a compiler flag like ``-DLOG_LEVEL=...``
- changing the default log level value in the ``Logging.h`` file as I suggested.

The library as such does assume there is no need to get debug output from **within** the library any more, so the default is set to ``LOG_LEVEL_ERROR``. 

BTW: if the default level is set with one of the methods above, you may change the verbosity **globally** at runtime by assigning a value to ``MBUlogLvl``, but only up to the given global default level. So you may compile everything with maximum verbosity for the cost of larger binaries and the need to explicitly restrict the level with ``MBUlogLvl`` once you do not need the verbose output any more.

**Reply by Andrew-Garage** [reply]
It works! I can even change the logging level from the computer, which can be useful in case of remote diagnostics. Or something else. In any case, I am glad that there is an opportunity to add such a feature. Thanks again!
![изображение](https://user-images.githubusercontent.com/89803784/216547024-a06898ef-1bdc-4a7a-bf7e-be336f2154f9.png)

**Reply by Miq1** [reply]
You are welcome! 😉
